src/mw.ts

```diff
@@ -0,0 +1,129 @@
+import { Context, Middleware } from '@curveball/core';
+import * as cookie from 'cookie';
+import MemoryStore from './memorystore';
+import { SessionOptions, SessionStore, SessionValues } from './types';
+import { randomBytes } from 'crypto';
+import { CsrfError } from './errors';
+
+/**
+ * This function returns a middleware function.
+ */
+export default function(options: SessionOptions): Middleware {
+
+  const cookieName = options.cookieName ? options.cookieName : 'CB';
+
+  const cookieOptions = options.cookieOptions || {
+    path: '/',
+    sameSite: 'lax',
+    httpOnly: true,
+  };
+
+  let store: SessionStore;
+
+  if (options.store === 'memory') {
+    store = new MemoryStore();
+  } else {
+    store = options.store;
+  }
+
+  /**
+   * Expire after 1 hour by default.
+   */
+  const expiry = options.expiry !== undefined ? options.expiry : 3600;
+
+  return async (ctx, next) => {
+
+    let sessionId = getSessionId(ctx, cookieName);
+    let sessionValues: SessionValues | null;
+
+    ctx.session = {};
+    ctx.sessionId = null;
+    ctx.getCsrf = async() => {
+      if (ctx.session['csrf-token']) {
+        return ctx.session['csrf-token'];
+      }
+      const bytes = await randomBytes(32);
+      const token = bytes.toString('base64');
+      ctx.session['csrf-token'] = token;
+      return token;
+    };
+    ctx.validateCsrf = (token?: string) => {
+      if (!token) {
+        token = ctx.request?.body?.['csrf-token'];
+      }
+      if (!token) {
+        throw new CsrfError('No CSRF token was found in the request body');
+      }
```

Comment:
i'm guessing the reason why there are 2 `if` statement is because even after the 1st statement, it didn't set token, and then throw an error?

Reply:
Yes, you can either pass a token to the function, but if it's not passed we'll check the body for a token.